From 754f493ef44d4a0f7b59e8c95e50b911631ac60b Mon Sep 17 00:00:00 2001
From: ysl2 <www.songli.yu@gmail.com>
Date: Sun, 29 Oct 2023 15:40:44 +0800
Subject: [PATCH] Bulk kill: arg.ui == 0 for normal kill current client; arg.ui
 == 1 for kill other clients in current tag (except current focusing client);
 arg.ui == 2 for kill all clients in current tag (include focusing client).

---
 bulkill.c    | 36 ++++++++++++++++++++++++++++++++++++
 config.def.h |  3 +++
 2 files changed, 39 insertions(+)
 create mode 100644 bulkill.c

diff --git a/bulkill.c b/bulkill.c
new file mode 100644
index 0000000..c73958f
--- /dev/null
+++ b/bulkill.c
@@ -0,0 +1,36 @@
+static void killthis(Window w);
+static void bulkill(const Arg *arg);
+
+void
+killthis(Window w) {
+	if (!sendevent(w, wmatom[WMDelete], NoEventMask, wmatom[WMDelete], CurrentTime, 0, 0, 0)) {
+		XGrabServer(dpy);
+		XSetErrorHandler(xerrordummy);
+		XSetCloseDownMode(dpy, DestroyAll);
+		XKillClient(dpy, w);
+		XSync(dpy, False);
+		XSetErrorHandler(xerror);
+		XUngrabServer(dpy);
+	}
+}
+
+void
+bulkill(const Arg *arg)
+{
+    Client *c;
+
+	if (!selmon->sel)
+		return;
+
+    if (!arg->ui || arg->ui == 0) {
+        killthis(selmon->sel->win);
+        return;
+    }
+
+    for (c = selmon->clients; c; c = c->next) {
+        if (!ISVISIBLE(c) || (arg->ui == 1 && c == selmon->sel))
+            continue;
+        killthis(c->win);
+    }
+}
+
diff --git a/config.def.h b/config.def.h
index 750529d..4d4bcef 100644
--- a/config.def.h
+++ b/config.def.h
@@ -65,6 +65,7 @@ static char dmenumon[2] = "0"; /* component of dmenucmd, manipulated in spawn()
 static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont, "-nb", col_gray1, "-nf", col_gray3, "-sb", col_cyan, "-sf", col_gray4, NULL };
 static const char *termcmd[]  = { "st", NULL };
 
+#include "bulkill.c"
 static const Key keys[] = {
 	/* modifier                     key        function        argument */
 	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
@@ -79,6 +80,8 @@ static const Key keys[] = {
 	{ MODKEY,                       XK_Return, zoom,           {0} },
 	{ MODKEY,                       XK_Tab,    view,           {0} },
 	{ MODKEY|ShiftMask,             XK_c,      killclient,     {0} },
+    { MODKEY|ControlMask,           XK_c,      bulkill,        {.ui = 1} },  // kill unselect
+    { MODKEY|ShiftMask|ControlMask, XK_c,      bulkill,        {.ui = 2} },  // killall
 	{ MODKEY,                       XK_t,      setlayout,      {.v = &layouts[0]} },
 	{ MODKEY,                       XK_f,      setlayout,      {.v = &layouts[1]} },
 	{ MODKEY,                       XK_m,      setlayout,      {.v = &layouts[2]} },
-- 
2.20.1

